CLDR-17851 SupplementalCalendarData parser, update ExampleGenerator

srl295 opened the pull request:
- minimal, modern parser
- minimal test case
- update example generator

CLDR-17851 

- Fixes original root cause of #4629 

- [ ] This PR completes the ticket.

<!--
Thank you for your pull request.
Please see https://cldr.unicode.org/index/process for general
information on contributing to CLDR.

1. Make sure the ticket is filed at
https://unicode-org.atlassian.net/projects/CLDR/
2. Update the PR title and first line of this
message to include the ticket ID (CLDR-_____)
3. You will be automatically asked to sign the contributors’
license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/cldr
- license: https://www.unicode.org/copyright.html#License
-->

ALLOW_MANY_COMMITS=true

srl295 commented:
@sffc this gives us a data driven approach that should be less crashy on era data changes

sffc reviewed (commented):
I don't understand enough about how CLDR works to be able to give this a thorough review

srl295 commented:
> I don't understand enough about how CLDR works to be able to give this a thorough review

mostly just a FYI that CLDR *could* now have data quality checks on `<calendarData>` (nothing, sadly, parsed it before)

srl295 commented:
example generator issue, will bring back soon

srl295 commented:
alright, BH now shows `AH 0` as its example.  Which is better than `AH 1` i guess.

anyway, everyone gets examples now

![image](https://github.com/user-attachments/assets/9bbc6cd4-1295-4cf3-8f1b-1e6c6161d5d0)

AEApple commented:
Thank you for doing this! I'm not familiar enough either to review this unfortunately, although I could try applying it in my branch and check if it fixed the error if that works?

srl295 commented:
Already did and it does work

El El vie, 25 abr 2025 a la(s) 6:24 p.m., Annemarie Apple <
***@***.***> escribió:

> *AEApple* left a comment (unicode-org/cldr#4642)
> <https://github.com/unicode-org/cldr/pull/4642#issuecomment-2831596602>
>
> Thank you for doing this! I'm not familiar enough either to review this
> unfortunately, although I could try applying it in my branch and check if
> it fixed the error if that works?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/cldr/pull/4642#issuecomment-2831596602>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGQZMZJVOIWBLKJ7VMBPMT23K73ZAVCNFSM6AAAAAB34QIRLKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZRGU4TMNRQGI>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,229 @@
+package org.unicode.cldr.util;
+
+import com.ibm.icu.impl.locale.XCldrStub.ImmutableMap;
```
This is wrong; we should never have XCldrStub in any CLDR code.

srl295 replied:
done

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,229 @@
+        private final XPathValue xpath;
+
+        EraData(final XPathValue xpath) {
+            this.xpath = xpath;
```
In general, for sup. data we have parsed out the data from each path and cached it in a friendly data structure, rather than parsing on every single call to access data.

macchiati replied:
If you don't have time to do those changes now, please file a ticket to fix in the future.

srl295 replied:
It actually doesn't re parse, the only overhead is a hash table look up on the frozen Xpath.

srl295 replied:
I was trying to keep the parser  simple as a model for others that need to be added

srl295 replied:
done

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,229 @@
+            return l;
+        }
+
+        /** only works within the same cal system */
```
This is then too fragile; add a calendar ID to the EraData, then compare first by the calendarID, then by the era num ber.

macchiati replied:
BTW, ComparisonChain is less error-prone.

srl295 replied:
done

macchiati reviewed (changes requested):
A few items

srl295 commented on tools/cldr-code/src/main/java/org/unicode/cldr/test/ExampleGenerator.java:
```diff
@@ -3151,30 +3100,87 @@ private void handleDateRangePattern(String value, List<String> examples) {
+        GregorianCalendar sampleDate = null;
+
+        if (startCal != null && endCal != null) {
+            // roll back a day to not hit the edge
```
done

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,269 @@
+        }
+
+        public String[] getAliases() {
+            final String s = xpath.getAttributeValue(INDEX, LDMLConstants.ALIASES);
```
Didn't notice this earlier. That's an awkward way to get an empty string array. Also, we should avoid String::split.

Use a static Splitter with splitToList, and trimming, and change the return value to a List

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,269 @@
+            type = Integer.parseInt(xpath.getAttributeValue(INDEX, LDMLConstants.TYPE));
+
+            start = xpath.getAttributeValue(INDEX, LDMLConstants.START);
+            startCalendar = forDateString(start);
```
A Calendar is pretty heavy-weight for this. Would be simplest to just extract the millis.

macchiati replied:
Actually, an even better reason is that then you (and your clients) can avoid complications with null. The earliest start date is Integer.MIN_VALUE and the latest end date is Integer.MAX_VALUE.

There is one further complication. Normally missing start/end dates get the above values. However, for the Japanese calendar, the end date is set based on the **next** line's start data - 1.
So
```
              <era type="0" start="645-6-19"/>
              <era type="1" start="650-2-15"/>
...
              <era type="235" start="1989-1-8" code="heisei"/>
              <era type="236" start="2019-5-1" code="reiwa"/>
```
Is equivalent to 
```
              <era type="0" start="645-6-19" end="650-2-14" />
              <era type="1" start="650-2-15" end="671-12-31"/>
...
              <era type="235" start="1989-1-8" end="2019-4-30" code="heisei"/>
              <era type="236" start="2019-5-1" code="reiwa"/>
```
So that means you don't know the end data for an era until the next one gets added. And that's assuming that you read them in file order. Otherwise you need to read them in, and sort by start date.

srl295 replied:
do we need more clarity in the spec here? 

before this, i couldn't find any code in CLDR that read these lines.

srl295 replied:
> So that means you don't know the end data for an era until the next one gets added. And that's assuming that you read them in file order. Otherwise you need to read them in, and sort by start date.

That's why i had the end time be null there.

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,269 @@
+import org.unicode.cldr.icu.LDMLConstants;
+
+public class SupplementalCalendarData implements Iterable<String> {
+    /** an <era> element */
```
```suggestion
    /** an <era> element, which represents the era information from a calendar such as:
    * <pre>         
    * <calendarData><supplementalData>
    *    <calendar type="islamic">
    *      <calendarSystem type="lunar" />
    *      <eras>
    *          <era type="0" start="622-7-15" code="islamic" aliases="ah"/>
    *           <era type="1" end="622-7-14" code="islamic-inverse" aliases="bh"/>
    *      </eras>
    *   </calendar>
    * </pre>
    */
```

srl295 replied:
or just link to the spec?

macchiati replied:
The behavior of start/end is documented in https://www.unicode.org/reports/tr35/tr35-69/tr35-dates.html#Calendar_Data

(It could use some wordsmithing to clarify that "open-ended" applies when there isn't a series; probably start that first. I'll file a ticket for that.)

macchiati replied:
Ticket: https://unicode-org.atlassian.net/browse/CLDR-18576

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,269 @@
+        }
+
+        /** only works within the same cal system */
+        @Override
```
```suggestion
        @Override
        /* Normally comparisons require all of the fields to be listed, to ensure that no 
         * two unequal values a and b will never have a.compareTo(b) = 0.
         * However, the structure of the survey tool guaranties that no two eras (for different calendarTypes)
         * will overlap in start/end dates.
         */
```

srl295 replied:
the survey tool doesn't guarantee this - do you mean the spec? the unit tests ought to validate this. But I'm not sure what all of the constraints are.

macchiati replied:
Right, we don't make that clear. Filed https://unicode-org.atlassian.net/browse/CLDR-18576

macchiati replied:
You're right. It is the spec not the Survey Tool.

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/SupplementalCalendarData.java:
```diff
@@ -0,0 +1,269 @@
+        public int compareTo(EraData o) {
+            return ComparisonChain.start()
+                    .compare(calendarType, o.calendarType)
+                    .compare(getLatest(), o.getLatest())
+                    .compare(getType(), o.getType())
```
```suggestion
                    .compare(getEnd(), o.getEnd())
```
Because of the above feature, you will never need the type for comparison; just the calendarType and end.

srl295 commented:
@macchiati this is getting pretty heavy weight. -and not working. Could I go back to the lightweight parser that succeeds for purposes of examples and split out the heavier api questions?

srl295 commented:
> @macchiati this is getting pretty heavy weight. -and not working. Could I go back to the lightweight parser that succeeds for purposes of examples and split out the heavier api questions?

Specifically, back to Friday's version (but memoize it so the object is only created once).  And split out a PR for the more heavyweight processor that tries to sort out the situation for Japanese, etc.

I was trying to make a parser that just reflected what was in the XML, to enable unit testing, but it is kind of turning into a general purpose implementation of the algorithm.

btangmu commented on tools/cldr-code/src/main/java/org/unicode/cldr/test/ExampleGenerator.java:
```diff
@@ -3151,30 +3100,87 @@ private void handleDateRangePattern(String value, List<String> examples) {
             }
+        } else {
+            // System.err.println("No good date for " + eraData);
+            // TODO: should be an error in TestSupplementalDataInfo
```
@srl295 does this TODO comment indicate more to be done on this ticket?
